packages/cli/src/databases/repositories/execution.repository.ts

```diff
@@ -399,21 +405,89 @@ export class ExecutionRepository extends Repository<ExecutionEntity> {
 
 		if (!executions.length) {
 			if (deleteConditions.ids) {
-				LoggerProxy.error('Failed to delete an execution due to insufficient permissions', {
+				Logger.error('Failed to delete an execution due to insufficient permissions', {
 					executionIds: deleteConditions.ids,
 				});
 			}
 			return;
 		}
 
 		const executionIds = executions.map(({ id }) => id);
-		const binaryDataManager = BinaryDataManager.getInstance();
-		await binaryDataManager.deleteBinaryDataByExecutionIds(executionIds);
-
 		do {
 			// Delete in batches to avoid "SQLITE_ERROR: Expression tree is too large (maximum depth 1000)" error
-			const batch = executionIds.splice(0, 500);
-			await this.delete(batch);
+			const batch = executionIds.splice(0, this.deletionBatchSize);
+			await this.softDelete(batch);
 		} while (executionIds.length > 0);
 	}
+
+	async pruneBySoftDeleting() {
+		Logger.verbose('Pruning (soft-deleting) execution data from database');
+
+		const maxAge = config.getEnv('executions.pruneDataMaxAge'); // in h
+		const maxCount = config.getEnv('executions.pruneDataMaxCount');
+
+		// Find ids of all executions that were stopped longer that pruneDataMaxAge ago
+		const date = new Date();
+		date.setHours(date.getHours() - maxAge);
+
+		const toPrune: Array<FindOptionsWhere<IExecutionFlattedDb>> = [
+			// date reformatting needed - see https://github.com/typeorm/typeorm/issues/2286
+			{ stoppedAt: LessThanOrEqual(DateUtils.mixedDateToUtcDatetimeString(date)) },
+		];
+
+		if (maxCount > 0) {
+			const executions = await this.find({
+				select: ['id'],
+				skip: maxCount,
+				take: 1,
+				order: { id: 'DESC' },
+			});
+
+			if (executions[0]) {
+				toPrune.push({ id: LessThanOrEqual(executions[0].id) });
+			}
+		}
+
+		const executionIds = (
+			await this.find({
+				select: ['id'],
+				where: toPrune,
+				take: this.deletionBatchSize,
+			})
+		).map(({ id }) => id);
+		await this.softDelete(executionIds);
```

Comment:
We can actually replace this code with a single update, where we set deletedAt = now() where stoppedAt <= "{cutDate}" - would be much more performant.

Reply:
Like the above comment, I definitely agree (and I mentioned this optimization to Adi [here](https://github.com/n8n-io/n8n/pull/7065#discussion_r1312869061)) but I wonder if this falls under improving existing pruning logic and so maybe to be tackled separately later rather than now? 

From what I see, typeorm supports neither batch updates nor returning IDs for some DBs.

For example, on this query...

```ts
const [timeBasedWhere, countBasedWhere] = toPrune;

await this.createQueryBuilder()
	.update(ExecutionEntity)
	.set({ deletedAt: new Date() })
	.where(
		new Brackets((qb) =>
			countBasedWhere
				? qb.where(timeBasedWhere).orWhere(countBasedWhere)
				: qb.where(timeBasedWhere),
		),
	)	
	.execute();
```

Adding `.limit(this.deletionBatchSize)` to this query triggers:

```
LimitOnUpdateNotSupportedError: Your database does not support LIMIT on UPDATE statements.
```

And adding `.returning(['id'])` triggers:

```
ReturningStatementNotSupportedError: OUTPUT or RETURNING clause only supported by Microsoft SQL Server or PostgreSQL or MariaDB databases.
```

Maybe an optimization to explore in future?

Reply:
Discussed with Omar, with this change batching for soft deletions is no longer needed - it was needed before to prevent a long list of IDs from triggering `SQLITE_ERROR: Expression tree is too large (maximum depth 1000)`

Reply:
Addressed here: [259fe75](https://github.com/n8n-io/n8n/pull/7092/commits/259fe75b8608754e3c41337dca022dbddbb1dd50)